codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/protocols/parse/JsonParserGenerator.kt

```diff
@@ -208,8 +236,29 @@ class JsonParserGenerator(
             rustBlock("match key.to_unescaped()?.as_ref()") {
                 for (member in members) {
                     rustBlock("${jsonName(member).dq()} =>") {
-                        withBlock("builder = builder.${member.setterName()}(", ");") {
-                            deserializeMember(member)
+                        when (codegenTarget) {
+                            CodegenTarget.CLIENT -> {
+                                withBlock("builder = builder.${member.setterName()}(", ");") {
+                                    deserializeMember(member)
+                                }
+                            }
+                            CodegenTarget.SERVER -> {
+                                if (symbolProvider.toSymbol(member).isOptional()) {
+                                    withBlock("builder = builder.${member.setterName()}(", ");") {
+                                        deserializeMember(member)
+                                    }
+                                } else {
+                                    rust("if let Some(v) = ")
+                                    deserializeMember(member)
+                                    rust(
+                                        """
+                                        {
```

Comment:
Nit: I think this could be made more readable if `deserializeMember` is refactored to return a `Writable`.